# **APA - Data Preparation Template**
## **DATA 3300**

## **Name(s):**

### **Read and follow these assignment instructions carefully! Ordinarily, we’d jump into cleansing data as needed and we’d each do it slightly differently. This is difficult to grade, so please go in the order of this document and follow these instructions.**

**For this assignment students should submit one .ipynb file, one clean, sorted Excel file, and a PDF of this iPython notebook file. See the assignment background for attribute descriptions and the assignment requirements link on the Canvas assignment page.**

**Students should alter code in lines with "replace code..." language. Otherwise, students do not need to alter the provided code.**

## **Q1**

**What does the term “data quality” refer to and why is it important within the context of using data to solve business problems?**

###  The degree to which a dataset can be efficiently and effectively processed and used. It is important because it helps create analyses that are meaningful, trustworthy, and that do not induce mayhem.

##**Q2**

**The lecture and textbook discussed five characteristics of data quality: accuracy, completeness, consistency, timeliness, and uniqueness. While “accuracy” in this dataset is hard for us to gauge without further inquiry and“timeliness” kind of depends on what it is the data are to be used for, there are clear examples of problems with each of the other three characteristics.**

**For each of the other three (completeness, consistency, uniqueness), identify a specific situation within the *hbdata-orig.csv* dataset (identify record numbers when applicable). Hint: For uniqueness, look at the VisitSpan attribute. Be sure your answers are clearly labelled and described.**

**This can be done by viewing the data in Excel or importing it into Python. If using Python, begin by importing your libraries and then read in the .csv file.**

In [1]:
import pandas as pd
import numpy as np

#just run this code block after importing libraries above, no need to alter the below code!
from datetime import datetime #you'll also need this method to convert a variable to date-time later in the assignment
import warnings
warnings.filterwarnings("ignore") #this can be removed but will help the code be a little cleaner by ignoring warnings

In [2]:
#replace with code for importing the dataset
df = pd.read_excel('/content/hbdata-orig24.xlsx')
#run this code block after importing dataset above
pd.set_option("display.max_rows", None, "display.max_columns", None) #View full dataframe
df

VisitID             BirdName  BirdType BirdSize  Gender  \
0       2203             Rose Kim     Rufus        S  female   
1       2204       Horace Prefect  Calliope        M    male   
2       2205   Edeltraud McKnight    Rufous        M  female   
3       2206     Herbert McKnight    Rufous        M    male   
4       2207    Walpurga Schwalbe    Anna's        M  female   
5       2208    Walpurga Schwalbe    Anna's        M  female   
6       2209                  NaN       NaN      NaN     NaN   
7       2210          Blossom Kim    Rufous        M  female   
8       2211          Gale Sharma    Rufous       XL  female   
9       2212        King Schwalbe    Anna's        L    male   
10      2213            Lucky Kim    Rufous        M    male   
11      2214        Ruby Schwalbe    Anna's        S  female   
12      2215          Norm Sharma    Rufous        M    male   
13      2216   Petunia O'Flaherty    Anna's        S  female   
14      2217         Sol Schwalbe    Anna's        M    male   
15      2218         Mitz Johnson     Rufus        M  female   
16      2219     Francine Johnson    Rufous        M  female   
17      2220        June McKnight    Rufous        M  female   
18      2221            Lucky Kim    Rufous        M    male   
19      2222        King Schwalbe    Anna's        L    male   
20      2223      Virgil Schwalbe    Anna's        M    male   
21      2224      Bort O'Flaherty    Anna's        L    male   
22      2225     Malachi Schwalbe    Anna's       XL    male   
23      2226        June McKnight    Rufous        M  female   
24      2227        Bobert Sharma    Rufous        M    male   
25      2228   Petunia O'Flaherty    Anna's        S  female   
26      2229        Ruby Schwalbe    Anna's        S  female   
27      2230   Edeltraud McKnight    Rufous        M  female   
28      2231        King Schwalbe    Anna's        L    male   
29      2232    Walpurga Schwalbe    Anna's        M  female   
30      2233        Renee Prefect  Calliope        M  female   
31      2234        June McKnight    Rufous        M  female   
32      2235   Edeltraud McKnight    Rufous        M  female   
33      2236   Petunia O'Flaherty    Anna's        S  female   
34      2237      Clancy McKnight    Rufous        M    male   
35      2238      Bort O'Flaherty    Anna's        L    male   
36      2239     Francine Johnson       NaN        M  female   
37      2240        Ruby Schwalbe    Anna's        S  female   
38      2241      Bort O'Flaherty    Anna's        L    male   
39      2242            Lucky Kim     Rufus        M    male   
40      2243        King Schwalbe    Anna's        L    male   
41      2244               Horace   Prefect      NaN      no   
42      2245         Mitz Johnson    Rufous        M  female   
43      2246        Renee Prefect  Calliope        M  female   
44      2247         Wally Sharma    Rufous        M    male   
45      2248         Mitz Johnson    Rufous        M  female   
46      2249        Renee Prefect  Calliope        M  female   
47      2250   Edeltraud McKnight    Rufous   medium  female   
48      2251        Renee Prefect  Calliope   medium  female   
49      2252      Judith McKnight    Rufous   medium  female   
50      2253          Blossom Kim    Rufous   medium  female   
51      2254         Cake Prefect  Calliope   medium    male   
52      2255      Virgil Schwalbe       NaN   medium    male   
53      2256       Horace Prefect  Calliope   medium    male   
54      2257      Judith McKnight    Rufous   medium  female   
55      2258     Herbert McKnight    Rufous        M    male   
56      2259          Gale Sharma    Rufous       XL  female   
57      2260     Hortence Prefect  Calliope        M  female   
58      2261         Wally Sharma    Rufous        M    male   
59      2262     Hortence Prefect  Calliope        M  female   
60      2263       Kevin McKnight    Rufous        M    male   
61      2264      Bill O'Flaherty    


*   Completeness: VisitIDs 2209, 2395, missing data
*   Consistency: VisitIDs 2250-2257, medium instead of M. Lots of oz instead of ml as well
*   Uniqueness: 2235-2336: Time is the same, duplicate entry

## **Q3**

**Answer the following questions regarding the importance of each of the named data scrubbing steps.**

### **3A**
**What negative outcome would likely occur during data analysis if duplicated visits were not removed (consolidated)?**

The dulicate data would be recorded multiple times, scewing the results of a particular bird or feeder. We would likely find that there are more birds observed than there should be, and their results would skew what we assume about that bird.

### **3B**

**What negative outcome would likely occur during data analysis if Location entries were not standardized?**

Some locations would have missing entries, for example "Deck. 2" would not be counted in "Deck 2"

### **3C**

**What negative outcome would likely occur during data analysis if, due to measurement issues, amount eaten values were not consistent?**

One might assume a bird has eaten much more or much less than it actually has. This can cause future predictions of how much to feed to be wrong.

## **Q4**

**In scrubbing the data, you should address each of the following. It’s recommended to go through this assignment in the order of steps below.**

### 4A

In Python trim all text columns to ensure there are no leading or trailing spaces on any of the data values. If you still see two seemingly identical entries, remember that trimming does not trim internal spaces between two words.

In [3]:
df_obj = df.select_dtypes(['object']) #select all columns containing text (of type object)
df[df_obj.columns] = df_obj.apply(lambda x: x.str.strip()) #trim all selected columns
#replace with code to view the header of the dataframe (cheat sheet!)
df.head()
#keep below code to verify date/time of your work
from datetime import datetime
now = datetime.now()
formatted_time = now.strftime("%Y-%m-%d %H:%M:%S")
print(f"Current Date and Time: {formatted_time}")

Current Date and Time: 2024-01-25 20:34:27


### 4B

Ensure each bird’s first and last name should be included in separate fields (columns).

In [4]:
#replace with code to split column, updating column names and delimiter type (cheat sheet)
df[['BirdFirstName','BirdLastName']] = df['BirdName'].str.split(' ', expand=True)
#replace with code to drop birdname column
df = df.drop('BirdName', axis = 1)
#replace with code to show df heading
df.head()

VisitID  BirdType BirdSize  Gender VisitDate                  VisitSpan  \
0     2203     Rufus        S  female       NaN  23:13:05.736-23:13:21.109   
1     2204  Calliope        M    male       NaN    07:45:07.65-07:45:17.97   
2     2205    Rufous        M  female       NaN    08:09:52.10-08:10:22.74   
3     2206    Rufous        M    male       NaN    08:41:54.07-08:42:07.48   
4     2207    Anna's        M  female       NaN    09:18:59.33-09:19:13.41   

     Location Hoverfeed Amount Unnamed: 10 BirdFirstName BirdLastName  
0  Back Patio       yes  17 ml         NaN          Rose          Kim  
1   NE Corner       yes   6 ml         NaN        Horace      Prefect  
2      Deck 2        no   5 ml         NaN     Edeltraud     McKnight  
3      Deck 1       yes   2 ml         NaN       Herbert     McKnight  
4      Garage       yes   3 ml         NaN      Walpurga     Schwalbe

### 4C
Ensure that data values are standardized in each column (pay particular attention to the textual columns, BirdName, BirdType, BirdSize, Gender, Location, and Hoverfeed) so that each “idea” (e.g., a species of bird) is specified (i.e., spelled) the same way.

This can be checked for each column by using the .value_counts() method! After applying the standarization, show that the correction has been made by using the .value_counts() method to display all categories for each column of interest.

In [5]:
#Replace with code to examine levels in BirdType using the value_counts() method, update column name
df['BirdType'].value_counts()

Rufous      103
Anna's       85
Calliope     61
Rufus        18
Prefect      14
Name: BirdType, dtype: int64

In [6]:
#Replace with code to replace redundant value in BirdType, for only 1 original and new label
df = df.replace('Rufus','Rufous')

In [7]:
#Replace with code to examine levels of BirdSize. Ignore null, "medium" and 20ml lines; these will be fixed later.
df['BirdSize'].value_counts()

M         162
S          48
L          33
XL         23
medium     13
20 ml       5
Name: BirdSize, dtype: int64

In [8]:
#Replace with code to replace redundant value in BirdSize
df = df.replace('medium','M')

In [9]:
#Replace with code to examine levels of Location. Ignore two lines with all numerical output; these will be fixed later.
df['Location'].value_counts()

Back Patio                 68
Deck 2                     54
Deck 1                     46
NE Corner                  44
Garage                     29
Entry                      29
Back  Patio                 8
Deck. 2                     7
15:16:40.19-15:16:55.98     4
13:14:33.36-13:15:09.25     2
Name: Location, dtype: int64

In [10]:
#Replace with code to replace redundant values of Location, for 2 original and new labels
df = df.replace({'Back  Patio':'Back Patio','Deck. 2':'Deck 2'})

### 4D
The amount eaten is very important to our analysis. Where this value is null (blank), omit the record. Note that values of 0 are not the same as null. Retain (don’t delete) the records where the value of Amount is 0. Other issues with this attribute will be addressed later.

Before removing missing columns first examine the length of the dataset currently using the len() method. Then split the *Amount* column into two columns, *Amount_Eat* and *Unit*. Finally, change *Amount_Eat* to a float object using pd.to_numeric().

Once missing values are dropped, view the length of the dataset again using the len() method to see how many observations were dropped.

In [11]:
len(df) #return length of dataframe (i.e., number of rows), just run this code block

291

In [12]:
#Replace with code to create new split columns for 'Amount_Eat' and 'Unit' from the 'Amount' column, updating column names and delimiter type
df[['Amount_Eat','Unit']] = df['Amount'].str.split(' ', expand=True)
#Replace with code to drop 'Amount' column
df = df.drop('Amount', axis = 1)
#Replace with code to show header of dataframe
df.head()

VisitID  BirdType BirdSize  Gender VisitDate                  VisitSpan  \
0     2203    Rufous        S  female       NaN  23:13:05.736-23:13:21.109   
1     2204  Calliope        M    male       NaN    07:45:07.65-07:45:17.97   
2     2205    Rufous        M  female       NaN    08:09:52.10-08:10:22.74   
3     2206    Rufous        M    male       NaN    08:41:54.07-08:42:07.48   
4     2207    Anna's        M  female       NaN    09:18:59.33-09:19:13.41   

     Location Hoverfeed Unnamed: 10 BirdFirstName BirdLastName Amount_Eat Unit  
0  Back Patio       yes         NaN          Rose          Kim         17   ml  
1   NE Corner       yes         NaN        Horace      Prefect          6   ml  
2      Deck 2        no         NaN     Edeltraud     McKnight          5   ml  
3      Deck 1       yes         NaN       Herbert     McKnight          2   ml  
4      Garage       yes         NaN      Walpurga     Schwalbe          3   ml

In [13]:
df['Amount_Eat'] = df['Amount_Eat'].apply(pd.to_numeric, errors='coerce') #convert Amount_Eat column variable dtype from object to float
#just run this code block
df

VisitID  BirdType BirdSize  Gender VisitDate                  VisitSpan  \
0       2203    Rufous        S  female       NaN  23:13:05.736-23:13:21.109   
1       2204  Calliope        M    male       NaN    07:45:07.65-07:45:17.97   
2       2205    Rufous        M  female       NaN    08:09:52.10-08:10:22.74   
3       2206    Rufous        M    male       NaN    08:41:54.07-08:42:07.48   
4       2207    Anna's        M  female       NaN    09:18:59.33-09:19:13.41   
5       2208    Anna's        M  female       NaN    09:28:41.64-09:28:52.64   
6       2209       NaN      NaN     NaN       NaN    10:56:26.40-10:56:45.32   
7       2210    Rufous        M  female       NaN    11:27:05.53-11:27:31.41   
8       2211    Rufous       XL  female       NaN    12:14:11.70-12:14:35.72   
9       2212    Anna's        L    male       NaN    12:14:45.98-12:15:14.45   
10      2213    Rufous        M    male       NaN    13:07:55.82-13:08:11.09   
11      2214    Anna's        S  female       NaN    14:08:35.90-14:08:41.95   
12      2215    Rufous        M    male       NaN    15:15:15.55-15:15:35.11   
13      2216    Anna's        S  female       NaN    15:38:48.39-15:39:11.49   
14      2217    Anna's        M    male       NaN    16:58:39.17-16:58:57.62   
15      2218    Rufous        M  female       NaN    17:25:12.74-17:25:34.44   
16      2219    Rufous        M  female       NaN    07:29:33.89-07:29:56.56   
17      2220    Rufous        M  female       NaN    09:52:51.78-09:53:14.73   
18      2221    Rufous        M    male       NaN    10:25:08.10-10:25:29.68   
19      2222    Anna's        L    male       NaN    11:09:34.78-11:10:01.07   
20      2223    Anna's        M    male       NaN    11:35:55.06-11:36:16.76   
21      2224    Anna's        L    male       NaN    11:54:20.11-11:54:29.64   
22      2225    Anna's       XL    male       NaN    12:26:37.48-12:27:05.68   
23      2226    Rufous        M  female       NaN    12:26:58.01-12:27:19.08   
24      2227    Rufous        M    male       NaN    12:34:38.96-12:34:56.59   
25      2228    Anna's        S  female       NaN    12:36:58.28-12:37:26.58   
26      2229    Anna's        S  female       NaN    12:37:00.80-12:37:17.85   
27      2230    Rufous        M  female       NaN    13:18:55.66-13:19:09.19   
28      2231    Anna's        L    male       NaN    13:23:11.98-13:23:29.77   
29      2232    Anna's        M  female       NaN    15:07:05.17-15:07:31.63   
30      2233  Calliope        M  female       NaN    15:54:15.27-15:54:41.58   
31      2234    Rufous        M  female       NaN    16:18:40.38-16:19:01.38   
32      2235    Rufous        M  female       NaN    16:33:19.74-16:33:30.57   
33      2236    Anna's        S  female       NaN    17:01:37.83-17:02:00.03   
34      2237    Rufous        M    male       NaN    17:28:37.12-17:28:54.44   
35      2238    Anna's        L    male       NaN    17:59:52.71-18:00:22.84   
36      2239       NaN        M  female       NaN    09:02:56.80-09:03:28.74   
37      2240    Anna's        S  female       NaN    11:16:29.94-11:16:52.88   
38      2241    Anna's        L    male       NaN    11:47:10.08-11:47:29.87   
39      2242    Rufous        M    male       NaN    12:02:30.95-12:03:00.41   
40      2243    Anna's        L    male       NaN    12:10:23.42-12:10:43.25   
41      2244   Prefect      NaN      no  Calliope                        NaN   
42      2245    Rufous        M  female       NaN    13:33:08.79-13:33:26.04   
43      2246  Calliope        M  female       NaN    13:53:55.26-13:54:22.18   
44      2247    Rufous        M    male       NaN    14:26:32.64-14:26:47.91   
45      2248    Rufous        M  female       NaN    14:53:47.69-14:54:18.06   
46      2249  Calliope        M  female       NaN    15:01:13.77-15:01:39.94   
47      2250    Rufous        M  female       NaN    17:40:07.86-17:40:33.43   
48      2251  Calliope        M  female       NaN    08:35:49.12-08:36:07.42   
49   

In [14]:
#Replace with code to remove missing data for the 'Amount_Eat' column
df = df[df['Amount_Eat'].notna()] # drop rows with missing data based on a specified column
len(df) #check the new length of the dataset

283

### 4E

Ensure that each visit is unique. In other words, no visit should be recorded more than once (duplicated). Remove duplicates from the VisitSpan column  using the drop_duplicates() method.

Then use the len() method to show how many observations were dropped.

In [15]:
df = df.drop_duplicates(subset=['VisitSpan']) #drop duplicate records from the 'VisitSpan' column

#Replace with code to check the new length of the dataset (see above)
len(df)

279

### 4F
Split the VisitSpan column into two columns: rename them StartTime and EndTime.

Then ensure that both time columns are set as datetime objects by using the pd.to_datetime() method.

In [16]:
#Replace with code to split 'VisitSpan' based on the '-' delimiter, create a StartTime and EndTime column
df[['StartTime', 'EndTime']] = df['VisitSpan'].str.split('-', expand = True) # split based on _ delimeter
df['StartTime'] = pd.to_datetime(df['StartTime'], format='%H:%M:%S.%f') #converts StartTime to hours:minutes:seconds format
df['EndTime'] = pd.to_datetime(df['EndTime'], format='%H:%M:%S.%f') #converts EndTime to hours:minutes:seconds format

#Replace with code to drop the 'VisitSpan' column
df = df.drop('VisitSpan', axis = 1)
#Replace with code to show header of dataframte
df.head()

VisitID  BirdType BirdSize  Gender VisitDate    Location Hoverfeed  \
0     2203    Rufous        S  female       NaN  Back Patio       yes   
1     2204  Calliope        M    male       NaN   NE Corner       yes   
2     2205    Rufous        M  female       NaN      Deck 2        no   
3     2206    Rufous        M    male       NaN      Deck 1       yes   
4     2207    Anna's        M  female       NaN      Garage       yes   

  Unnamed: 10 BirdFirstName BirdLastName  Amount_Eat Unit  \
0         NaN          Rose          Kim        17.0   ml   
1         NaN        Horace      Prefect         6.0   ml   
2         NaN     Edeltraud     McKnight         5.0   ml   
3         NaN       Herbert     McKnight         2.0   ml   
4         NaN      Walpurga     Schwalbe         3.0   ml   

                StartTime                 EndTime  
0 1900-01-01 23:13:05.736 1900-01-01 23:13:21.109  
1 1900-01-01 07:45:07.650 1900-01-01 07:45:17.970  
2 1900-01-01 08:09:52.100 1900-01-01 08:10:22.740  
3 1900-01-01 08:41:54.070 1900-01-01 08:42:07.480  
4 1900-01-01 09:18:59.330 1900-01-01 09:19:13.410

### 4G
Some recordings of Amount appear to have been made in ounces rather than in milliliters. All of them should be in milliliters. Convert those in ounces to milliliters (1 oz. = 29.57 ml).

Start by subsetting the data to only those observations containing oz, then apply the conversion to the *Amount_Eat* column and replace the values in the *Unit* column with ml.


In [17]:
df_oz = df[df['Unit'] == 'oz'] #subsets dataset to only contain oz
df_oz['Amount_Eat'] = 29.57*df['Amount_Eat'] #applies conversion formula
df_oz['Unit'] = df_oz['Unit'].replace('oz', 'ml') #replaces oz with ml
df_oz.head() #just run this entire code block as is

VisitID BirdType BirdSize  Gender VisitDate    Location Hoverfeed  \
33      2236   Anna's        S  female       NaN  Back Patio        no   
36      2239      NaN        M  female       NaN      Deck 1       yes   
39      2242   Rufous        M    male       NaN      Deck 2       yes   
99      2302   Rufous        L  female       NaN  Back Patio       yes   
105     2308   Anna's        L    male       NaN   NE Corner        no   

    Unnamed: 10 BirdFirstName BirdLastName  Amount_Eat Unit  \
33          NaN       Petunia   O'Flaherty     0.00000   ml   
36          NaN      Francine      Johnson     7.92476   ml   
39          NaN         Lucky          Kim     0.14785   ml   
99          NaN          Rose          Kim    12.47854   ml   
105         NaN          King     Schwalbe     6.38712   ml   

                  StartTime                 EndTime  
33  1900-01-01 17:01:37.830 1900-01-01 17:02:00.030  
36  1900-01-01 09:02:56.800 1900-01-01 09:03:28.740  
39  1900-01-01 12:02:30.950 1900-01-01 12:03:00.410  
99  1900-01-01 12:54:14.400 1900-01-01 12:54:35.650  
105 1900-01-01 19:12:52.240 1900-01-01 19:13:18.720

###**4H**

**Next, you'll combine your original dataframe with the new subset dataframe using the concat() method. Finally, drop rows containing 'oz' in the *Unit* column.**

**Make sure you have the correct number of observations still by using the len() method.**

In [18]:
frames = [df, df_oz]
df = pd.concat(frames) #combines the original df with the new subsetted df
df = df[df.Unit != 'oz'] #removes rows containing oz

#Replace with code to examine length of dataframe (see earlier portions of this assignment)
len(df)

279

### 4I

Createa new column, called “IsFemale”. This column should include a 1 if the hummingbird that made the visit is female, or 0 if the hummingbird is male.

In [19]:
#Replace with code to create new column called IsFemale using the np.where method
df['IsFemale'] = np.where(df['Gender'] == 'female',1,0)
df.head()

VisitID  BirdType BirdSize  Gender VisitDate    Location Hoverfeed  \
0     2203    Rufous        S  female       NaN  Back Patio       yes   
1     2204  Calliope        M    male       NaN   NE Corner       yes   
2     2205    Rufous        M  female       NaN      Deck 2        no   
3     2206    Rufous        M    male       NaN      Deck 1       yes   
4     2207    Anna's        M  female       NaN      Garage       yes   

  Unnamed: 10 BirdFirstName BirdLastName  Amount_Eat Unit  \
0         NaN          Rose          Kim        17.0   ml   
1         NaN        Horace      Prefect         6.0   ml   
2         NaN     Edeltraud     McKnight         5.0   ml   
3         NaN       Herbert     McKnight         2.0   ml   
4         NaN      Walpurga     Schwalbe         3.0   ml   

                StartTime                 EndTime  IsFemale  
0 1900-01-01 23:13:05.736 1900-01-01 23:13:21.109         1  
1 1900-01-01 07:45:07.650 1900-01-01 07:45:17.970         0  
2 1900-01-01 08:09:52.100 1900-01-01 08:10:22.740         1  
3 1900-01-01 08:41:54.070 1900-01-01 08:42:07.480         0  
4 1900-01-01 09:18:59.330 1900-01-01 09:19:13.410         1

### 4J
Create a new column called Duration, which includes the duration of the visit in seconds. Creat this column from the *StartTime* and *EndTime* columns then convert to seconds using dt.total_seconds().

In [20]:
df['Duration'] = df['EndTime'] - df['StartTime'] #creates new 'Duration' column by subtracting StartTime from EndTime

In [21]:
df['Duration'] = df['Duration'].dt.total_seconds() #converts 'Duration' to seconds
df.head()

VisitID  BirdType BirdSize  Gender VisitDate    Location Hoverfeed  \
0     2203    Rufous        S  female       NaN  Back Patio       yes   
1     2204  Calliope        M    male       NaN   NE Corner       yes   
2     2205    Rufous        M  female       NaN      Deck 2        no   
3     2206    Rufous        M    male       NaN      Deck 1       yes   
4     2207    Anna's        M  female       NaN      Garage       yes   

  Unnamed: 10 BirdFirstName BirdLastName  Amount_Eat Unit  \
0         NaN          Rose          Kim        17.0   ml   
1         NaN        Horace      Prefect         6.0   ml   
2         NaN     Edeltraud     McKnight         5.0   ml   
3         NaN       Herbert     McKnight         2.0   ml   
4         NaN      Walpurga     Schwalbe         3.0   ml   

                StartTime                 EndTime  IsFemale  Duration  
0 1900-01-01 23:13:05.736 1900-01-01 23:13:21.109         1    15.373  
1 1900-01-01 07:45:07.650 1900-01-01 07:45:17.970         0    10.320  
2 1900-01-01 08:09:52.100 1900-01-01 08:10:22.740         1    30.640  
3 1900-01-01 08:41:54.070 1900-01-01 08:42:07.480         0    13.410  
4 1900-01-01 09:18:59.330 1900-01-01 09:19:13.410         1    14.080

### 4K

View the full dataset, to spot check your cleaned data and address any remaining issues you notice.

In [22]:
pd.set_option("display.max_rows", None, "display.max_columns", None) #views full dataframe
df

VisitID  BirdType BirdSize  Gender VisitDate                 Location  \
0       2203    Rufous        S  female       NaN               Back Patio   
1       2204  Calliope        M    male       NaN                NE Corner   
2       2205    Rufous        M  female       NaN                   Deck 2   
3       2206    Rufous        M    male       NaN                   Deck 1   
4       2207    Anna's        M  female       NaN                   Garage   
5       2208    Anna's        M  female       NaN                   Deck 2   
7       2210    Rufous        M  female       NaN                   Deck 1   
8       2211    Rufous       XL  female       NaN               Back Patio   
9       2212    Anna's        L    male       NaN                NE Corner   
10      2213    Rufous        M    male       NaN                   Deck 2   
11      2214    Anna's        S  female       NaN                   Deck 2   
12      2215    Rufous        M    male       NaN               Back Patio   
13      2216    Anna's        S  female       NaN                   Deck 2   
14      2217    Anna's        M    male       NaN               Back Patio   
15      2218    Rufous        M  female       NaN                   Deck 2   
16      2219    Rufous        M  female       NaN                   Deck 1   
17      2220    Rufous        M  female       NaN               Back Patio   
18      2221    Rufous        M    male       NaN               Back Patio   
19      2222    Anna's        L    male       NaN                NE Corner   
20      2223    Anna's        M    male       NaN                    Entry   
21      2224    Anna's        L    male       NaN                   Deck 2   
22      2225    Anna's       XL    male       NaN               Back Patio   
23      2226    Rufous        M  female       NaN               Back Patio   
24      2227    Rufous        M    male       NaN                   Deck 1   
25      2228    Anna's        S  female       NaN                   Garage   
26      2229    Anna's        S  female       NaN                NE Corner   
27      2230    Rufous        M  female       NaN                   Deck 1   
28      2231    Anna's        L    male       NaN                NE Corner   
29      2232    Anna's        M  female       NaN                NE Corner   
30      2233  Calliope        M  female       NaN                NE Corner   
31      2234    Rufous        M  female       NaN               Back Patio   
32      2235    Rufous        M  female       NaN                   Deck 2   
34      2237    Rufous        M    male       NaN                   Deck 2   
35      2238    Anna's        L    male       NaN                NE Corner   
37      2240    Anna's        S  female       NaN                NE Corner   
38      2241    Anna's        L    male       NaN                NE Corner   
40      2243    Anna's        L    male       NaN                    Entry   
42      2245    Rufous        M  female       NaN               Back Patio   
43      2246  Calliope        M  female       NaN               Back Patio   
44      2247    Rufous        M    male       NaN               Back Patio   
45      2248    Rufous        M  female       NaN                   Deck 2   
46      2249  Calliope        M  female       NaN                    Entry   
47      2250    Rufous        M  female       NaN                   Deck 1   
48      2251  Calliope        M  female       NaN                   Deck 2   
49      2252    Rufous        M  female       NaN                   Deck 2   
50      2253    Rufous        M  female       NaN               Back Patio   
51      2254  Calliope        M    male       NaN                   Garage   
52      2255       NaN        M    male       NaN                   Deck 2   
53      2256  Calliope        M    male       NaN                   Garage   
54      2257    Rufous        M  female       NaN               Back Patio   
55      2258    Rufous     

In [23]:
#Replace with code to remove any remaining unwanted columns
df = df.drop('Unnamed: 10', axis = 1)
df.head()

VisitID  BirdType BirdSize  Gender VisitDate    Location Hoverfeed  \
0     2203    Rufous        S  female       NaN  Back Patio       yes   
1     2204  Calliope        M    male       NaN   NE Corner       yes   
2     2205    Rufous        M  female       NaN      Deck 2        no   
3     2206    Rufous        M    male       NaN      Deck 1       yes   
4     2207    Anna's        M  female       NaN      Garage       yes   

  BirdFirstName BirdLastName  Amount_Eat Unit               StartTime  \
0          Rose          Kim        17.0   ml 1900-01-01 23:13:05.736   
1        Horace      Prefect         6.0   ml 1900-01-01 07:45:07.650   
2     Edeltraud     McKnight         5.0   ml 1900-01-01 08:09:52.100   
3       Herbert     McKnight         2.0   ml 1900-01-01 08:41:54.070   
4      Walpurga     Schwalbe         3.0   ml 1900-01-01 09:18:59.330   

                  EndTime  IsFemale  Duration  
0 1900-01-01 23:13:21.109         1    15.373  
1 1900-01-01 07:45:17.970         0    10.320  
2 1900-01-01 08:10:22.740         1    30.640  
3 1900-01-01 08:42:07.480         0    13.410  
4 1900-01-01 09:19:13.410         1    14.080

In [24]:
pd.set_option("display.max_rows", None, "display.max_columns", None) #views full dataframe
df

VisitID  BirdType BirdSize  Gender VisitDate                 Location  \
0       2203    Rufous        S  female       NaN               Back Patio   
1       2204  Calliope        M    male       NaN                NE Corner   
2       2205    Rufous        M  female       NaN                   Deck 2   
3       2206    Rufous        M    male       NaN                   Deck 1   
4       2207    Anna's        M  female       NaN                   Garage   
5       2208    Anna's        M  female       NaN                   Deck 2   
7       2210    Rufous        M  female       NaN                   Deck 1   
8       2211    Rufous       XL  female       NaN               Back Patio   
9       2212    Anna's        L    male       NaN                NE Corner   
10      2213    Rufous        M    male       NaN                   Deck 2   
11      2214    Anna's        S  female       NaN                   Deck 2   
12      2215    Rufous        M    male       NaN               Back Patio   
13      2216    Anna's        S  female       NaN                   Deck 2   
14      2217    Anna's        M    male       NaN               Back Patio   
15      2218    Rufous        M  female       NaN                   Deck 2   
16      2219    Rufous        M  female       NaN                   Deck 1   
17      2220    Rufous        M  female       NaN               Back Patio   
18      2221    Rufous        M    male       NaN               Back Patio   
19      2222    Anna's        L    male       NaN                NE Corner   
20      2223    Anna's        M    male       NaN                    Entry   
21      2224    Anna's        L    male       NaN                   Deck 2   
22      2225    Anna's       XL    male       NaN               Back Patio   
23      2226    Rufous        M  female       NaN               Back Patio   
24      2227    Rufous        M    male       NaN                   Deck 1   
25      2228    Anna's        S  female       NaN                   Garage   
26      2229    Anna's        S  female       NaN                NE Corner   
27      2230    Rufous        M  female       NaN                   Deck 1   
28      2231    Anna's        L    male       NaN                NE Corner   
29      2232    Anna's        M  female       NaN                NE Corner   
30      2233  Calliope        M  female       NaN                NE Corner   
31      2234    Rufous        M  female       NaN               Back Patio   
32      2235    Rufous        M  female       NaN                   Deck 2   
34      2237    Rufous        M    male       NaN                   Deck 2   
35      2238    Anna's        L    male       NaN                NE Corner   
37      2240    Anna's        S  female       NaN                NE Corner   
38      2241    Anna's        L    male       NaN                NE Corner   
40      2243    Anna's        L    male       NaN                    Entry   
42      2245    Rufous        M  female       NaN               Back Patio   
43      2246  Calliope        M  female       NaN               Back Patio   
44      2247    Rufous        M    male       NaN               Back Patio   
45      2248    Rufous        M  female       NaN                   Deck 2   
46      2249  Calliope        M  female       NaN                    Entry   
47      2250    Rufous        M  female       NaN                   Deck 1   
48      2251  Calliope        M  female       NaN                   Deck 2   
49      2252    Rufous        M  female       NaN                   Deck 2   
50      2253    Rufous        M  female       NaN               Back Patio   
51      2254  Calliope        M    male       NaN                   Garage   
52      2255       NaN        M    male       NaN                   Deck 2   
53      2256  Calliope        M    male       NaN                   Garage   
54      2257    Rufous        M  female       NaN               Back Patio   
55      2258    Rufous     

## Q5

When you’re finished cleaning the datafile, sort the file by Visit ID (low to high), then save the cleaned file as an Excel document for submission via Canvas.

In [25]:
#Replace with code to set index to VisitID
df = df.set_index('VisitID')
df.sort_index() #sorts index

BirdType BirdSize  Gender VisitDate                 Location  \
VisitID                                                                 
2203       Rufous        S  female       NaN               Back Patio   
2204     Calliope        M    male       NaN                NE Corner   
2205       Rufous        M  female       NaN                   Deck 2   
2206       Rufous        M    male       NaN                   Deck 1   
2207       Anna's        M  female       NaN                   Garage   
2208       Anna's        M  female       NaN                   Deck 2   
2210       Rufous        M  female       NaN                   Deck 1   
2211       Rufous       XL  female       NaN               Back Patio   
2212       Anna's        L    male       NaN                NE Corner   
2213       Rufous        M    male       NaN                   Deck 2   
2214       Anna's        S  female       NaN                   Deck 2   
2215       Rufous        M    male       NaN               Back Patio   
2216       Anna's        S  female       NaN                   Deck 2   
2217       Anna's        M    male       NaN               Back Patio   
2218       Rufous        M  female       NaN                   Deck 2   
2219       Rufous        M  female       NaN                   Deck 1   
2220       Rufous        M  female       NaN               Back Patio   
2221       Rufous        M    male       NaN               Back Patio   
2222       Anna's        L    male       NaN                NE Corner   
2223       Anna's        M    male       NaN                    Entry   
2224       Anna's        L    male       NaN                   Deck 2   
2225       Anna's       XL    male       NaN               Back Patio   
2226       Rufous        M  female       NaN               Back Patio   
2227       Rufous        M    male       NaN                   Deck 1   
2228       Anna's        S  female       NaN                   Garage   
2229       Anna's        S  female       NaN                NE Corner   
2230       Rufous        M  female       NaN                   Deck 1   
2231       Anna's        L    male       NaN                NE Corner   
2232       Anna's        M  female       NaN                NE Corner   
2233     Calliope        M  female       NaN                NE Corner   
2234       Rufous        M  female       NaN               Back Patio   
2235       Rufous        M  female       NaN                   Deck 2   
2236       Anna's        S  female       NaN               Back Patio   
2237       Rufous        M    male       NaN                   Deck 2   
2238       Anna's        L    male       NaN                NE Corner   
2239          NaN        M  female       NaN                   Deck 1   
2240       Anna's        S  female       NaN                NE Corner   
2241       Anna's        L    male       NaN                NE Corner   
2242       Rufous        M    male       NaN                   Deck 2   
2243       Anna's        L    male       NaN                    Entry   
2245       Rufous        M  female       NaN               Back Patio   
2246     Calliope        M  female       NaN               Back Patio   
2247       Rufous        M    male       NaN               Back Patio   
2248       Rufous        M  female       NaN                   Deck 2   
2249     Calliope        M  female       NaN                    Entry   
2250       Rufous        M  female       NaN                   Deck 1   
2251     Calliope        M  female       NaN                   Deck 2   
2252       Rufous        M  female       NaN                   Deck 2   
2253       Rufous        M  female       NaN               Back Patio   
2254     Calliope        M    male       NaN                   Garage   
2255          NaN        M    male       NaN                   Deck 2   
2256     Calliope        M    male       NaN                   Garage   
2257       Rufous        M  female       NaN               Back Pat

In [26]:
df

BirdType BirdSize  Gender VisitDate                 Location  \
VisitID                                                                 
2203       Rufous        S  female       NaN               Back Patio   
2204     Calliope        M    male       NaN                NE Corner   
2205       Rufous        M  female       NaN                   Deck 2   
2206       Rufous        M    male       NaN                   Deck 1   
2207       Anna's        M  female       NaN                   Garage   
2208       Anna's        M  female       NaN                   Deck 2   
2210       Rufous        M  female       NaN                   Deck 1   
2211       Rufous       XL  female       NaN               Back Patio   
2212       Anna's        L    male       NaN                NE Corner   
2213       Rufous        M    male       NaN                   Deck 2   
2214       Anna's        S  female       NaN                   Deck 2   
2215       Rufous        M    male       NaN               Back Patio   
2216       Anna's        S  female       NaN                   Deck 2   
2217       Anna's        M    male       NaN               Back Patio   
2218       Rufous        M  female       NaN                   Deck 2   
2219       Rufous        M  female       NaN                   Deck 1   
2220       Rufous        M  female       NaN               Back Patio   
2221       Rufous        M    male       NaN               Back Patio   
2222       Anna's        L    male       NaN                NE Corner   
2223       Anna's        M    male       NaN                    Entry   
2224       Anna's        L    male       NaN                   Deck 2   
2225       Anna's       XL    male       NaN               Back Patio   
2226       Rufous        M  female       NaN               Back Patio   
2227       Rufous        M    male       NaN                   Deck 1   
2228       Anna's        S  female       NaN                   Garage   
2229       Anna's        S  female       NaN                NE Corner   
2230       Rufous        M  female       NaN                   Deck 1   
2231       Anna's        L    male       NaN                NE Corner   
2232       Anna's        M  female       NaN                NE Corner   
2233     Calliope        M  female       NaN                NE Corner   
2234       Rufous        M  female       NaN               Back Patio   
2235       Rufous        M  female       NaN                   Deck 2   
2237       Rufous        M    male       NaN                   Deck 2   
2238       Anna's        L    male       NaN                NE Corner   
2240       Anna's        S  female       NaN                NE Corner   
2241       Anna's        L    male       NaN                NE Corner   
2243       Anna's        L    male       NaN                    Entry   
2245       Rufous        M  female       NaN               Back Patio   
2246     Calliope        M  female       NaN               Back Patio   
2247       Rufous        M    male       NaN               Back Patio   
2248       Rufous        M  female       NaN                   Deck 2   
2249     Calliope        M  female       NaN                    Entry   
2250       Rufous        M  female       NaN                   Deck 1   
2251     Calliope        M  female       NaN                   Deck 2   
2252       Rufous        M  female       NaN                   Deck 2   
2253       Rufous        M  female       NaN               Back Patio   
2254     Calliope        M    male       NaN                   Garage   
2255          NaN        M    male       NaN                   Deck 2   
2256     Calliope        M    male       NaN                   Garage   
2257       Rufous        M  female       NaN               Back Patio   
2258       Rufous        M    male       NaN                   Deck 1   
2259       Rufous       XL  female       NaN                   Deck 1   
2260     Calliope        M  female       NaN                   Gara

In [27]:
#export your cleaned Excel file for submission, along with your notebook

#For Google Colab
from google.colab import files
df.to_excel("cleaned_data_HB.xlsx")
files.download("cleaned_data_HB.xlsx")

#For JupyterNotebooks
df.to_excel("/content/hbdata-orig24.xlsx")

#keep below code to verify date/time of your work
from datetime import datetime
now = datetime.now()
formatted_time = now.strftime("%Y-%m-%d %H:%M:%S")
print(f"Current Date and Time: {formatted_time}")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Current Date and Time: 2024-01-25 20:34:30


For submission to Canvas: Download/Save your .ipynb file, print your .ipynb file to PDF, and export your cleaned dataset as an Excel file.